envNext

friedolino78 opened the pull request:
GUI is on
https://github.com/mruby-zest/mruby-zest-build/pull/107

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added Bezier control points for envelope curve shaping, enabling more expressive amplitude/pitch envelopes.
  * Introduced new OSC commands for managing envelope control points.

* **Improvements**
  * Envelope playback now uses Bezier interpolation for smoother, more natural curves.
  * Enhanced release timing behavior for more accurate envelope control.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
add bezier interpolation to ENV

JohannesLorenz commented:
Rename this from "En vnext" to "Env next"?

fundamental commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -215,15 +245,13 @@ static const rtosc::Ports localPorts = {
-
-        for (int i=env->Penvpoints; i>=curpoint+1; i--) {
+        int i;
+        for (i=env->Penvpoints; i>=curpoint+1; i--) {
```
?

friedolino78 replied:
ok, to be removed

fundamental commented on src/Synth/Envelope.cpp:
```diff
@@ -133,12 +143,31 @@ void Envelope::watch(float time, float value)
     }
 }
 
+inline float lerp(float a, float b, float t)
```
defined, but not used

friedolino78 replied:
ok, will be removed, too

fundamental commented on src/Synth/Envelope.cpp:
```diff
@@ -187,11 +216,12 @@ float Envelope::envout(bool doWatch)
-    else
-        out = envval[currentpoint - 1]
-              + (envval[currentpoint] - envval[currentpoint - 1]) * t;
+    else { // if we have to interpolate between points
```
Am I correct in thinking this PR removes linear mode?

friedolino78 replied:
> Am I correct in thinking this PR removes linear mode?

if (envcpval == 0) the curve is linear

JohannesLorenz replied:
@fundamental Does the answer solve your concern?

fundamental commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -545,6 +577,7 @@ void EnvelopeParams::getfromXML(XMLwrapper& xml)
 
     if(!xml.hasparreal("A_dt")) {
         A_dt = dTREAL(xml.getpar127("A_dt", 0));
+        if (A_dt == 0.0f) A_dt= 0.0001f;
```
Why does this constraint only apply to attack

friedolino78 replied:
guess, I missed the others because they don't have 0 as default value. i will add them.

fundamental commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -263,9 +291,9 @@ EnvelopeParams::EnvelopeParams(unsigned char Penvstretch_,
+    A_dt  = 0.0001f;
     D_dt  = 0.009;
-    R_dt  = 0.009;
+    R_dt  = 0.499f;
```
IIRC these values are overwritten later, so is the change in the release needed?

friedolino78 replied:
fixed

fundamental commented:
As per the GUI comments, I'm not convinced that the currently proposed bezier behavior should be the final one.

friedolino78 commented:
> As per the GUI comments, I'm not convinced that the currently proposed bezier behavior should be the final one.

i changed it to 2nd order beziers with 2 control points per segment.

JohannesLorenz commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -545,8 +606,11 @@ void EnvelopeParams::getfromXML(XMLwrapper& xml)
 
     if(!xml.hasparreal("A_dt")) {
         A_dt = dTREAL(xml.getpar127("A_dt", 0));
+        if (A_dt == 0.0f) A_dt= 0.0001f;
```
These 3 should also go into a separate PR imo.

JohannesLorenz reviewed (commented):
For now, I gave only a few comments in order to clean up the PR a bit.

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -193,8 +194,58 @@ static const rtosc::Ports localPorts = {
+            }
+        }
+        rEnd},
+    //~ {"envcpx", rShort("bezier x") rDoc("Envelope Control Points"), NULL,
```
i tried to keep things simple by using only two parameters per segment, especially for gui simplicity.

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -274,8 +327,12 @@ EnvelopeParams::EnvelopeParams(unsigned char Penvstretch_,
+        envcpy[i*2+1] = 0.0f;
     }
-    envdt[0]        = 0.0f; //not used
+    envdt[0]        = 0.0001f; //not used
```
ok

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -193,8 +194,58 @@ static const rtosc::Ports localPorts = {
+
+        //~ }
+        //~ rEnd},
+    {"envcpy", rShort("bezier y") rDoc("Envelope Control Points"), NULL,
```
I'm not quite sure I understand. What would the default value do in an array? the mechanism is the same (copied) as for other arrays. What is missing?

friedolino78 replied:
done

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -274,8 +327,12 @@ EnvelopeParams::EnvelopeParams(unsigned char Penvstretch_,
     for(int i = 0; i < MAX_ENVELOPE_POINTS; ++i) {
         envdt[i]  = dTREAL(32);
         Penvval[i] = 64;
+        envcpx[i*2] = 0.0f;
```
good idea

JohannesLorenz commented:
@friedolino78 Let me know when you have addressed all the review comments. At least some seem not to be solved yet, so i will just wait for now.

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -193,8 +194,58 @@ static const rtosc::Ports localPorts = {
+        const int M = rtosc_narguments(msg);
+        if(M == 0) {
+            rtosc_arg_t args[N];
+            char arg_types[N+1] = {};
```
yes ;-)

Commit:
add unit to ENV time ports

Commit:
fix wrong bezier scaling at lin mode

Commit:
make MAX_ENVELOPE_CPOINTS 2 * MAX_ENVELOPE_POINTS

Commit:
fix the backward compatibility of the bezier function

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -122,7 +122,8 @@ static const rtosc::Ports localPorts = {
               rPreset(sub_freq, 0.254f),         rPreset(sub_bandwidth, 0.970f),
-              rDefault(0.0f),
+              rPreset(sub_filter, 0.127f),
+              rDefault(0.0001f),
```
ok

friedolino78 commented:
> @friedolino78 Let me know when you have addressed all the review comments. At least some seem not to be solved yet, so i will just wait for now.

I guess your comments are all solved.

Commit:
fix default values for A_dt

Commit:
make envcpy a parameter

Commit:
add breadcast to envcpy port

Commit:
remove parameter attribute from envcpy port

JohannesLorenz committed:
Add an envcpy blob port, making the other an alias

Commit:
remove broadcast from envcpy alias port

Commit:
remove unneeded assert header

JohannesLorenz reviewed (commented):
Partial review, to speed up progress.

JohannesLorenz commented:
Note: This whole branch needs a rebase, otherwise it cannot be compiled on a modern system (CMake compatibility).

JohannesLorenz reviewed (commented):
Complete code/style review.

I didn't test it yet.

friedolino78 commented on src/Params/EnvelopeParams.cpp:
```diff
@@ -238,8 +238,55 @@ static const rtosc::Ports localPorts = {
+
+            }
+            char part_loc[128];
+            strncpy(part_loc, d.loc, sizeof(part_loc));
```
ok

friedolino78 commented on src/Synth/Envelope.cpp:
```diff
@@ -168,7 +202,7 @@ float Envelope::envout(bool doWatch)
             out = envval[releaseindex];
         else
-            out = envoutval + (envval[releaseindex] - envoutval) * t; // linear interpolation envoutval and envval[releaseindex]
+            out = bezier3(envoutval, envcpy[releaseindex*2-1], envcpy[releaseindex*2], envval[releaseindex], t);
```
the control points have the index of the target value.
same as in:
`out = bezier3(envval[currentpoint - 1], envcpy[currentpoint*2-1], envcpy[currentpoint*2], envval[currentpoint], t);`

friedolino78 commented on src/Synth/Envelope.cpp:
```diff
@@ -76,6 +82,7 @@ Envelope::Envelope(EnvelopeParams &pars, float basefreq, float bufferdt,
     currentpoint = 1; //the envelope starts from 1
     keyreleased  = false;
     t = 0.0f;
+    tRelease = 0.0f;
```
ok

friedolino78 commented on src/Synth/Envelope.cpp:
```diff
@@ -59,15 +60,20 @@ Envelope::Envelope(EnvelopeParams &pars, float basefreq, float bufferdt,
                                  - 64.0f) / 64.0f) - 1.0f) * 100.0f;
                 if(pars.Penvval[i] < 64)
                     envval[i] = -envval[i];
+                envcpy[i] = 1000.0f * pars.envcpy[i];
```
yes and what matches the range of the values itself

Commit:
fixes due to review comments

friedolino78 commented on src/Synth/Envelope.cpp:
```diff
@@ -168,7 +194,7 @@ float Envelope::envout(bool doWatch)
             out = envval[releaseindex];
         else
-            out = envoutval + (envval[releaseindex] - envoutval) * t; // linear interpolation envoutval and envval[releaseindex]
+            out = bezier3(envoutval, envcpy[releaseindex*2-1], envcpy[releaseindex*2], envval[releaseindex], t);
```
you are right. it has to be initialized with 0. I will add it in the header.

Commit:
add initialization for envcpy.

JohannesLorenz commented on src/Synth/Envelope.cpp:
```diff
@@ -215,7 +249,7 @@ float Envelope::envout(bool doWatch)
 
     if(doWatch) {
-        watch(currentpoint + t, envoutval);
+        watch(currentpoint + t + tRelease, envoutval);
```
@fundamental Should this go into a separate PR for documentation, or are you fine with adding this to this PR here?

JohannesLorenz replied:
> depends if the rest of the envNext branch is ready

I will try to finish the review soon, then it should be ready.

friedolino78 replied:
https://github.com/zynaddsubfx/zynaddsubfx/pull/398

JohannesLorenz reviewed (approved):
Code approved.

Testing review should be next. However, I cannot do it because of mruby-zest/mruby-zest-build#127 :(

Commit:
fix bezier cp handling

JohannesLorenz commented:
@friedolino78 What was the state here? Was it fully working in Zest?

Commit:
adjust adnotetest for new ENV behaviour

friedolino78 commented:
> @friedolino78 What was the state here? Was it fully working in Zest?

it is working. I was using it quite a lot. but i noticed some quirks in the drag handling on some env points. (after sustain iirc). I guess that's to be solved in the GUI PR
